model/map/src/main/java/org/keycloak/models/map/authorization/MapResourceServerStore.java

```diff
@@ -0,0 +1,131 @@
+
+    public MapResourceServerStore(KeycloakSession session, MapStorage<String, MapResourceServerEntity, ResourceServer> resourceServerStore, AuthorizationProvider provider) {
+        this.resourceServerStore = resourceServerStore;
+        this.tx = resourceServerStore.createTransaction(session);
+        this.authorizationProvider = provider;
+        session.getTransactionManager().enlist(tx);
+    }
+
+    private MapResourceServerEntity registerEntityForChanges(MapResourceServerEntity origEntity) {
+        final MapResourceServerEntity res = tx.read(origEntity.getId(), id -> Serialization.from(origEntity));
+        tx.updateIfChanged(origEntity.getId(), res, MapResourceServerEntity::isUpdated);
+        return res;
+    }
+
+    private ResourceServer entityToAdapter(MapResourceServerEntity origEntity) {
+        if (origEntity == null) return null;
+        // Clone entity before returning back, to avoid giving away a reference to the live object to the caller
+        return new MapResourceServerAdapter(registerEntityForChanges(origEntity), authorizationProvider.getStoreFactory());
+    }
+
+    @Override
+    public ResourceServer create(String clientId) {
+        LOG.tracef("create(%s)%s", clientId, getShortStackTrace());
+        
+        if (clientId == null) return null;
+
+        if (!StorageId.isLocalStorage(clientId)) {
+            throw new ModelException("Creating resource server from federated ClientModel not supported");
+        }
+
+        if (tx.read(clientId) != null) {
+            throw new ModelDuplicateException("Resource server already exists: " + clientId);
+        }
+
+        MapResourceServerEntity entity = new MapResourceServerEntity(clientId);
+
+        tx.create(entity.getId(), entity);
+
+        return entityToAdapter(entity);
+    }
+
+    @Override
+    public void delete(String id) {
+        LOG.tracef("delete(%s, %s)%s", id, getShortStackTrace());
+        if (id == null) return;
+
+        PolicyStore policyStore = authorizationProvider.getStoreFactory().getPolicyStore();
+        policyStore.findByResourceServer(id).stream()
+            .map(Policy::getId)
+            .forEach(policyStore::delete);
+
+        PermissionTicketStore permissionTicketStore = authorizationProvider.getStoreFactory().getPermissionTicketStore();
+        permissionTicketStore.findByResourceServer(id).stream()
+                .map(PermissionTicket::getId)
+                .forEach(permissionTicketStore::delete);
+
+        ResourceStore resourceStore = authorizationProvider.getStoreFactory().getResourceStore();
+        resourceStore.findByResourceServer(id).stream()
+                .map(Resource::getId)
+                .forEach(resourceStore::delete);
+
+        ScopeStore scopeStore = authorizationProvider.getStoreFactory().getScopeStore();
+        scopeStore.findByResourceServer(id).stream()
+                .map(Scope::getId)
+                .forEach(scopeStore::delete);
```

Comment:
This is *very* inefficient and would easily cause transaction timeouts.
If this weren't for authz services, this would be required to be changed to `ModelCriteriaBuilder` and bulk delete.
Since it is authz and it is going to be handled later on by referential integrity, let's keep it as is for now.

Reply:
I agree. I will make sure to fix this place with the KEYCLOAK-17780 Jira.